### DISCUSSION TITLE
"One does not simply write an Eclipse plugin" AKA the lack of documentation for newcomers

### DISCUSSION THREAD

**Discussion by segevfiner** [eclipse-tycho/tycho#3481] [Q&A]
I wanted to explore what it would take to write an Eclipse plugin for the product in the company I work at (We already have a VS Code and IntelliJ plugins), I tried doing the obvious thing and simply downloading Eclipse and trying to create a plugin using the PDE wizards, but it felt like something is missing there, and reading around, and looking at the GitHub repos of core Eclipse plugins, it seems like stuff has shifted to using Tycho instead, with no indication or guidance as such in the Eclipse IDE itself, nor any obvious way to get started.

Any documentation seems strewn about random sites in the Internet, and it all feels rather incomplete, or expects that you have a lot of arcane knowledge in setting up Maven, p2, OSGi (Or maybe just be one of the guys who developed Eclipse in the first place) to actually figure out what anything does. Or maybe I just failed to find the proper docs? Doesn't seem that way though.

With the rant out of the way, I'll list a bunch of questions about stuff that I can't figure out what's the right way to do or do at all in the hopes that someone will help, and maybe also to point out what's missing from whatever documentation that does exist:

1. How are you supposed to create a new Eclipse plugin project? There is no obvious wizard, official arch type or anything like that to create a new project using Tycho, the wizards in Eclipse seem to use PDE/Build, which is documented as deprecated... and there seems to be a bunch of configuration that one would want to do for a proper setup that has no documentation either, such as, configuring a target platform, API Tools target?, and so on.

2. If PDE/Build is deprecated, how do you open a Maven/Tycho project in Eclipse, do you import it as a maven project? How do you run/debug it then?

3. How do you use dependencies? and publish a plugin using them? This isn't a run-of-the-mill Java classpath project, and I can't figure out what's the right way to use dependencies in either PDE/Build or Tycho, with multiple conflicting sources of information around the Internet, from copy Jars to this directory, to random Maven snippets that download Jars to a directory automatically, to assorted CLI commands, wizards, or Maven pom.xml to convert libraries to Eclipse plugins, to mysteriously adding them to the target platform, and publishing the plugin with them somehow working?!, And the whole OSGi vs non-OSGi thing... I have existing vanilla Java/Kotlin code I'd like to use which is built with Gradle and uses various libraries, no idea how to integrate such a thing into an Eclipse plugin.

4. How do you properly publish? There are random instructions around using either assorted Eclipse wizards, update site projects, or Tycho to build an "update site"/"repository", and I won't be surprised that some of those create legacy update sites... What's the up-to-date way? And further more, how do you publish updates to an existing such update site/repository? If you just FTP the thing it will overwrite the old versions, making them no longer available...

5. How would you handle native binaries/executables that you want to vendor and run from the plugin?

6. Can you use Kotlin in an Eclipse plugin? As we have plenty of existing Kotlin code we might want to use rather than convert to vanilla Java.. But I heard that Kotlin support in Eclipse is either bad or outright broken (The "official" plugin just breaks Eclipse)... Being a language developed by JetBrains and all, their focus is on their own IDEs after all... 

7. I remember the Eclipse browser used to use the system browser component, e.g. IE on Windows...But I'd heard there is a way to get it to use CEF instead (Though my experience with CEF was that it is buggy to the core). How would you configure a plugin to depend on and use that?

**Comment by laeubi** [comment]
> I tried doing the obvious thing and simply downloading Eclipse and trying to create a plugin using the PDE wizards, but it felt like something is missing there, and reading around, and looking at the GitHub repos of core Eclipse plugins, it seems like stuff has shifted to using Tycho instead, with no indication or guidance as such in the Eclipse IDE itself, nor any obvious way to get started.

It could help to explain what you tried, maybe even share the code in a public repository and link it here, but basically PDE == Development Time, Tycho == build time even though you can use other tools as well of course.

> Any documentation seems strewn about random sites in the Internet, and it all feels rather incomplete, or expects that you have a lot of arcane knowledge in setting up Maven, p2, OSGi (Or maybe just be one of the guys who developed Eclipse in the first place) to actually figure out what anything does. 

I think most assume that you have some basic knowledge yes but this one should give a quite good first introduction for newcomers: https://www.vogella.com/tutorials/eclipseplatform.html

> How are you supposed to create a new Eclipse plugin project?

- https://www.vogella.com/tutorials/EclipsePlugin/article.html#exercise-creating-and-using-a-first-plug-in

> There is no obvious wizard, official arch type or anything like that to create a new project using Tycho

The easiest is using a structured layout then Tycho will find everything itself usually:
- https://tycho.eclipseprojects.io/doc/latest/StructuredBuild.html

Also there are some demo projects around (even though they are very basic and could need polishing):
- https://github.com/eclipse-tycho/tycho/tree/main/demo

> The wizards in Eclipse seem to use PDE/Build, which is documented as deprecated

PDE is not deprecated, only the (ant) base pde-build tool, but you wont need that at all for first steps and can do everything from your IDE, e.g. if you develop an Eclipse Plugin you don't need a target platform at all, just use your running Eclipse as a target

> If PDE/Build is deprecated, how do you open a Maven/Tycho project in Eclipse, do you import it as a maven project? How do you run/debug it then?

You usually use PDE to run/debug and maven/tycho to build/publish it with your CI

> How do you use dependencies? and publish a plugin using them? This isn't a run-of-the-mill Java classpath project, and I can't figure out what's the right way to use dependencies in either PDE/Build or Tycho

It depends, you usually have two classes of dependencies:

1. Dependencies offered by the Eclipse Platform, you would need these as they provide you with the extensions you use to enhance Eclipse with a plugin
2. Third party stuff, there are several ways, the most convenient is probably creating a target platform and add some maven artifacts there

after that, you will import them in your manifest and can use them as you would for a regular jar based project

> I have existing vanilla Java/Kotlin code I'd like to use which is built with Gradle and uses various libraries, no idea how to integrate such a thing into an Eclipse plugin

I think that's a bit out of scope here but basically if you have a jar somewhere around you can use it.

> How do you properly publish?

There are several way and the answer depends on how you define "properly" but the most common way is publishing a P2 site what usually is  just an almost empty project with a `category.xml` where you define what should go in, then Tycho produces a site for you can upload to any http server to share it. There are other options but these are more about advanced use cases.

> How would you handle native binaries/executables that you want to vendor and run from the plugin?

I would suggest to postpone this topic unless you have setup a basic example but this usually only requires to add some metadata to your plugin and the rest is java as usual

> Can you use Kotlin in an Eclipse plugin? As we have plenty of existing Kotlin code we might want to use rather than convert to vanilla Java.. 

If you can setup the necessary toolchain and it produces a classfile in the end why not.

> But I heard that Kotlin support in Eclipse is either bad or outright broken (The "official" plugin just breaks Eclipse)... Being a language developed by JetBrains and all, their focus is on their own IDEs after all...

I can't tell about that because I don't know kotlin nor the plugins, but the developers of these plugin for sure desire feedback / enhancements from skilled users.

> I remember the Eclipse browser used to use the system browser component, e.g. IE on Windows...But I'd heard there is a way to get it to use CEF instead (Though my experience with CEF was that it is buggy to the core). How would you configure a plugin to depend on and use that?

Its hard to give advice based on this, basically, yes and yes and maybe :-)
But thats really another topic you should look into after solving 1..6


Finally if you need any dedicated support or training regarding Eclipse/Tycho feel free to contact me or see:
- https://github.com/eclipse-tycho/tycho?tab=readme-ov-file#getting-support

**Reply by segevfiner** [reply]
Thanks for answering! So I guess the tutorials in [vogella](https://www.vogella.com/) are the "official" docs for now? Let me try picking your brain a bit, or anyone else that might read this.

> > I tried doing the obvious thing and simply downloading Eclipse and trying to create a plugin using the PDE wizards, but it felt like something is missing there, and reading around, and looking at the GitHub repos of core Eclipse plugins, it seems like stuff has shifted to using Tycho instead, with no indication or guidance as such in the Eclipse IDE itself, nor any obvious way to get started.
> 
> It could help to explain what you tried, maybe even share the code in a public repository and link it here, but basically PDE == Development Time, Tycho == build time even though you can use other tools as well of course.
> 
What I'm confused about is that if I create a project using the PDE wizards and run it using PDE, won't this just use the PDE build stuff under the hood without Tycho involved? If I than add Tycho, it's basically kind of a Maven project, if I import the Maven pom into Eclipse (M2E and all that), what happens then with the PDE?

> > Any documentation seems strewn about random sites in the Internet, and it all feels rather incomplete, or expects that you have a lot of arcane knowledge in setting up Maven, p2, OSGi (Or maybe just be one of the guys who developed Eclipse in the first place) to actually figure out what anything does.
> 
> I think most assume that you have some basic knowledge yes but this one should give a quite good first introduction for newcomers: https://www.vogella.com/tutorials/eclipseplatform.html
> 
There are plenty of tutorials there, but not some overview of how p2, OSGi and all of those stuff fit together into the architecture of Eclipse, and how they effect you in developing a plugin isn't unless I missed it? But maybe that's just me liking to understand the internals of what I'm doing... rather than just going: "it works".

> > How are you supposed to create a new Eclipse plugin project?
> 
> * https://www.vogella.com/tutorials/EclipsePlugin/article.html#exercise-creating-and-using-a-first-plug-in
> 
> > There is no obvious wizard, official arch type or anything like that to create a new project using Tycho
> 
> The easiest is using a structured layout then Tycho will find everything itself usually:
> 
> * https://tycho.eclipseprojects.io/doc/latest/StructuredBuild.html
> 
> Also there are some demo projects around (even though they are very basic and could need polishing):
> 
> * https://github.com/eclipse-tycho/tycho/tree/main/demo
> 
So basically create a new project using the PDE wizards and add the pom.xml for the structured build? But should you then open the individiual projects in Eclipse or import the Maven pom.xml (M2E) to work on the project?

> > The wizards in Eclipse seem to use PDE/Build, which is documented as deprecated
> 
> PDE is not deprecated, only the (ant) base pde-build tool, but you wont need that at all for first steps and can do everything from your IDE, e.g. if you develop an Eclipse Plugin you don't need a target platform at all, just use your running Eclipse as a target
> > If PDE/Build is deprecated, how do you open a Maven/Tycho project in Eclipse, do you import it as a maven project? How do you run/debug it then?
> 
> You usually use PDE to run/debug and maven/tycho to build/publish it with your CI

I guess I'm confused about how does the IDE itself build the plugin when you open it as a project, doesn't that also use PDE/Build? If you open a Tycho project, can/will it use that instead?

> > How do you use dependencies? and publish a plugin using them? This isn't a run-of-the-mill Java classpath project, and I can't figure out what's the right way to use dependencies in either PDE/Build or Tycho
> 
> It depends, you usually have two classes of dependencies:
> 
> 1. Dependencies offered by the Eclipse Platform, you would need these as they provide you with the extensions you use to enhance Eclipse with a plugin
> 2. Third party stuff, there are several ways, the most convenient is probably creating a target platform and add some maven artifacts there
> 
> after that, you will import them in your manifest and can use them as you would for a regular jar based project

So I do need a target platform? As I'm trying to develop an Eclipse plugin that needs both dependencies developed my me/my company and third-party classic Maven ones (That might or might not be ready for OSGi). If I add a Maven dependency like that, how does packaging the final plugin work for someone to install it? As he will also need those dependencies, and they are obviously not a part of his Eclipse installation. They need to be vendored or installed alongside the plugin somehow...

> > I have existing vanilla Java/Kotlin code I'd like to use which is built with Gradle and uses various libraries, no idea how to integrate such a thing into an Eclipse plugin
> 
> I think that's a bit out of scope here but basically if you have a jar somewhere around you can use it.

I guess you can't integrate a Gradle build directly into a Maven or Eclipse PDE project more directly... The question is if I have an external project like that is how to set it up so that the jar will also be packaged correctly when I want to publish the plugin later on. I can try to depend on the relative build path of the Jar (In a monorepo) or publish it to the local Maven repository, but I'm not sure what's recommended for such a dependency...

> > How do you properly publish?
> 
> There are several way and the answer depends on how you define "properly" but the most common way is publishing a P2 site what usually is just an almost empty project with a `category.xml` where you define what should go in, then Tycho produces a site for you can upload to any http server to share it. There are other options but these are more about advanced use cases.

What's missing for me here is that you will get a new site that only has the latest version on each such build. Wouldn't you want to keep old versions around in the update site? If you just upload that you will overwrite the existing update site and will have only the latest version available.

> > How would you handle native binaries/executables that you want to vendor and run from the plugin?
> 
> I would suggest to postpone this topic unless you have setup a basic example but this usually only requires to add some metadata to your plugin and the rest is java as usual

Most of the core function of the plugin will come from those so that will be the next logical step after I got the scaffolding going. The question is mostly about packaging, I can just try to cram everything into the Jar and extract or if there is some way to publish an unpacked plugin with such binaries, but it might be nicer to get it to only download the binaries needed for the current platform directly as part of the plugin install from the update site (Rather than manually at plugin runtime). And I think there is support for platform specific parts, I'm just not sure what's the right way to configure that? Any examples/tutorials around?

> > Can you use Kotlin in an Eclipse plugin? As we have plenty of existing Kotlin code we might want to use rather than convert to vanilla Java..
> 
> If you can setup the necessary toolchain and it produces a classfile in the end why not.
> 
> > But I heard that Kotlin support in Eclipse is either bad or outright broken (The "official" plugin just breaks Eclipse)... Being a language developed by JetBrains and all, their focus is on their own IDEs after all...
> 
> I can't tell about that because I don't know kotlin nor the plugins, but the developers of these plugin for sure desire feedback / enhancements from skilled users.

I wonder if they do... from Googling, it seems like JetBrains is kinda unresponsive in the issue tracker to the point someone forked the plugin :P

> > I remember the Eclipse browser used to use the system browser component, e.g. IE on Windows...But I'd heard there is a way to get it to use CEF instead (Though my experience with CEF was that it is buggy to the core). How would you configure a plugin to depend on and use that?
> 
> Its hard to give advice based on this, basically, yes and yes and maybe :-) But thats really another topic you should look into after solving 1..6
> 
> Finally if you need any dedicated support or training regarding Eclipse/Tycho feel free to contact me or see:
> 
> * https://github.com/eclipse-tycho/tycho?tab=readme-ov-file#getting-support

Guess that's another thing to research once I have the other things in place, or ask somewhere else in the Eclipse ecosystem where someone that might be more familiar with this might know. I guess if it has an update site, it's just a matter of figuring out how to depend on plugins from external update sites or mirror those dependencies?

**Reply by laeubi** [reply]
> Thanks for answering! So I guess the tutorials in [vogella](https://www.vogella.com/) are the "official" docs for now?

No it is just a common place where many aspects are more handled in a tutorial like way, the official documentation is here:
https://help.eclipse.org/latest/nav/2

> What I'm confused about is that if I create a project using the PDE wizards and run it using PDE, won't this just use the PDE build stuff under the hood without Tycho involved? 

Basically yes

> If I than add Tycho, it's basically kind of a Maven project, if I import the Maven pom into Eclipse (M2E and all that), what happens then with the PDE?

You don't have to use a pom, Tycho is reading everything from PDE metadata, you usually only have a man/configurator pom in the root if you use [structured layouts](https://tycho.eclipseprojects.io/doc/latest/StructuredBuild.html), that's the main purpose of Tycho.

> So basically create a new project using the PDE wizards and add the pom.xml for the structured build? But should you then open the individiual projects in Eclipse or import the Maven pom.xml (M2E) to work on the project?

As mentioned you don't need a pom in your project, you usually will just import the project(s) into your workspace, of course still can use m2e for the root/configurator pom but that's not mandatory.

> I guess I'm confused about how does the IDE itself build the plugin when you open it as a project, doesn't that also use PDE/Build?

It uses "the Eclipse IDE" (or more formally JDT) to compile and PDE itself is a plugin to eclipse to support other things like running a product / launch a dev eclipse and so on

> If you open a Tycho project, can/will it use that instead?

There is no Tycho project, as tycho is a maven plugin, but you can also use maven with pde if you like but I would simply start with one thing at a time.

> So I do need a target platform?

It is useful for most cases beside basic examples, yes.

> As I'm trying to develop an Eclipse plugin that needs both dependencies developed my me/my company and third-party classic Maven ones (That might or might not be ready for OSGi).

m2e maven target locations allow adding missing OSGi metadata automatically for you

> If I add a Maven dependency like that, how does packaging the final plugin work for someone to install it? As he will also need those dependencies, and they are obviously not a part of his Eclipse installation. They need to be vendored or installed alongside the plugin somehow...

Dependencies will usually be included in your updatesite as well and included automatically when you install your plugin.

> What's missing for me here is that you will get a new site that only has the latest version on each such build. Wouldn't you want to keep old versions around in the update site? If you just upload that you will overwrite the existing update site and will have only the latest version available.

This all depends on your deployment strategies and development cycles, maybe even retention policy. It is usual for projects to have one "latest" url and some with the version in it for each release.

> I guess you can't integrate a Gradle build directly into a Maven or Eclipse PDE project more directly... 

It depends on how you define "integrate", as gradle is a build tool it usually makes no sense to mix concerns and try to have two competing ones but of course there are options I just won't recommend that for newcommers.

Theoretically Tycho could maybe work on as a Gradle plugin as well, its just that no one ever asked for it and it would be some efforts that will either require a contribution from someone or offering a funding for such feature.

> The question is mostly about packaging, I can just try to cram everything into the Jar and extract or if there is some way to publish an unpacked plugin with such binaries, but it might be nicer to get it to only download the binaries needed for the current platform directly as part of the plugin install from the update site

Eclipse support so called "native fragments" where you can include a native for a given os/arch and then it picks to install the one suitable for your current running environment, thats not that hard, but as said usually one should get the thing flying before going on with such tasks as it is often easier explained by example.

If you are curious you can for example look how Eclipse itself is doing it [here](https://github.com/eclipse-platform/eclipse.platform/tree/master/resources/bundles) 

> I wonder if they do... from Googling, it seems like JetBrains is kinda unresponsive in the issue tracker to the point someone forked the plugin

As said I'm not very familiar with this topic that why I'd rather stay at the things widely adopted/supported (plain java) but that might vary :-)

>  I guess if it has an update site, it's just a matter of figuring out how to depend on plugins from external update sites or mirror those dependencies?

In general yes.

**Reply by mickaelistria** [reply]
Unlike some other technologies, Eclipse plugin development has a clear separation between the development and the build task; they involve different technologies (which share some layers), but they are not exclusive.

> So I guess the tutorials in [vogella](https://www.vogella.com/) are the "official" docs for now?

Not the official tutorial of Eclipse Platform/PDE, but officially the best tutorial for Platform/PDE ;)

> What I'm confused about is that if I create a project using the PDE wizards and run it using PDE, won't this just use the PDE build stuff under the hood without Tycho involved? 

Not really, it uses internal logic of PDE, no full build in involved in running it with PDE. But you don't really need to care about that when developing usually.

> If I than add Tycho, it's basically kind of a Maven project, if I import the Maven pom into Eclipse (M2E and all that), what happens then with the PDE?

A project can be both PDE and Maven at the same time. However, Tycho has a pom-less extension so the project may not be detected as a Maven one if you don't want/need to add pom.

> But maybe that's just me liking to understand the internals of what I'm doing... rather than just going: "it works".

I think it's easier to start with "it works" and then discover how progressively.
OSGi provides some runtime dependency management with classpath isolation (allowing multiple versions of a lib to coexist in the same runtime), lifecycle handling (lazy loading/startup of bundles), and some service frameworks that are convenient to use in Eclipse RCP world.
p2 creates/processes OSGi dependency management metadata to make them available and used at *installation/provisioning*, so installing one artifact will install the right things. It's more or less the dnf or apt of Eclipse Plugins.
PDE and Tycho uses both: you use p2 (eg in target platforms) to define what is available for your product to consume, then you develop your bundles with OSGi and consuming what p2 made available as potential dependencies, and then you create some p2 repository or product with p2, which looks at your artifacts and "aggregates" it together with what is resolved as useful.

> So I do need a target platform? 

There is always a target platform involved. By default in the IDE, it's the current IDE. If it's good enough, just use it and don't bother further.
Tycho will require you to define either a target-platform or some (p2) repositories where to fetch dependencies. So you will quickly prefer to maintain such a .target file to control what can become visible to your build, both in the IDE and with Tycho. But as long as you don't feel the need for it, don't bother.

> I guess you can't integrate a Gradle build directly into a Maven or Eclipse PDE project more directly...

Not easily. Better giving up the idea of involving Gradle in your build at the moment.

> Wouldn't you want to keep old versions around in the update site?

We usually not. We have different p2 sites for different versions.

**Comment by segevfiner** [comment]
So I created the structure as described in https://tycho.eclipseprojects.io/doc/4.0.5/StructuredBuild.html, but it seems that Tycho requires a target platform or it fails to run, but I can't see any documentation for how to create a basic target platform for developing an Eclipse plugin agains't version X of Eclipse, the default of just using the running Eclipse isn't working when running from Maven/Tycho.

Also when opening the pom.xml from that doc in Eclipse, I'm getting an `'build.plugins.plugin.version' for org.eclipse.tycho:target-platform-configuration must be a valid version but is '${tycho-version}'.`, does M2E not resolve properties from `.mvn/maven.config`? Should this property be specified in the `pom.xml` instead?

I guess for some shared resource that is per plugin I can start a singleton from the activator, but is there anything for per project state?

Is there a more appropriate place for plugin development specific questions?

**Comment by laeubi** [comment]
> Also when opening the pom.xml from that doc in Eclipse, I'm getting an 'build.plugins.plugin.version' for org.eclipse.tycho:target-platform-configuration must be a valid version but is '${tycho-version}'., does M2E not resolve properties from .mvn/maven.config? Should this property be specified in the pom.xml instead?

m2e supports it but lemminx (the editor used by m2e) seems not, this is similar to:
- https://github.com/eclipse/lemminx-maven/issues/19

but you probably want to open a dedicated issue.

> it seems that Tycho requires a target platform or it fails to run, but I can't see any documentation for how to create a basic target platform for developing an Eclipse plugin agains't version X of Eclipse

See https://tycho.eclipseprojects.io/doc/latest/TargetPlatform.html

**Reply by segevfiner** [reply]
> > Also when opening the pom.xml from that doc in Eclipse, I'm getting an 'build.plugins.plugin.version' for org.eclipse.tycho:target-platform-configuration must be a valid version but is '${tycho-version}'., does M2E not resolve properties from .mvn/maven.config? Should this property be specified in the pom.xml instead?
> 
> m2e supports it but lemminx (the editor used by m2e) seems not, this is similar to:
> 
> * [Support Maven extensions.xml eclipse/lemminx-maven#19](https://github.com/eclipse/lemminx-maven/issues/19)
> 
> but you probably want to open a dedicated issue.
> 

Opened https://github.com/eclipse/lemminx-maven/issues/549
> > it seems that Tycho requires a target platform or it fails to run, but I can't see any documentation for how to create a basic target platform for developing an Eclipse plugin agains't version X of Eclipse
> 
> See https://tycho.eclipseprojects.io/doc/latest/TargetPlatform.html
It's not clear from that document what I should really do... Tycho is giving me:
```
target definition file '/Users/segevfiner/prj/swimm-eclipse/ide/extensions/eclipse/plugins/io.swimm.eclipse/../../target-platform.target' not found for project
```

Seems like the PDE can create target definitions, but it's not clear what I should put in it, as the default is simply just a reference to the running Eclipse install which Tycho doesn't understand. If I understand what this is meant to do, It feels like there should be some way to have a target definition that specifies some specific Eclipse EPP package? e.g. Eclipse for Java version xxxx-xx? But it's not obvious how to do so using the current PDE... or the docs I managed to find...

**Reply by laeubi** [reply]
You can use https://tycho.eclipseprojects.io/doc/latest/TargetPlatform.html#simple-target-platform-configuration you dont need a target file, just the URL of the current eclipse release (what is "the running platform"), technically you could use an exploded eclipse.zip file as well but thats rater unusual configuration.

**Reply by segevfiner** [reply]
Added a repository definition as specified to the POM and Tycho still complain with the same target definition error.

**Reply by laeubi** [reply]
> Added a repository definition as specified to the POM and Tycho still complain with the same target definition error.

You most probably have configured then somewhere to use that file, you should remove that ...

**Reply by segevfiner** [reply]
Oh the template from the structured builds article had that. Thanks!

Now if I want to use some Maven dependency? How do I do so? I read somewhere that there is a way to add them to a target definition and have it work in both Eclipse and Tycho and the final packaging of the plugin, but I don't have a target definition with that config, nor if that's really possible and the correct exact steps to do so.

**Reply by laeubi** [reply]
Yes the best is to create a target file, you will want there:

1. The Updatesite you used earlier most likely using the Eclipse SDK feature (probably more later)
2. add any maven dependency you need

then this can be used in Tycho and the IDE.

**Reply by segevfiner** [reply]
Argh. I did that. And also set it to include transitive Maven dependencies. And upon trying to use a package from the maven dependency I got random errors of missing packages/imports, had to restart Eclipse where it then started to complain about missing OSGi constraints in `io.gsonfire` and `javax.ws.rs-api` which if I understand how this works has been generated for them by the build process and likely has mistakes.

Not sure how to proceed to instruct it how to add those missing stuff to the wrapped packages or otherwise exclude them and bring proper OSGi versions of them somehow. I'm not sure if the maven integration for the target platform thing even supports excluding some transitive dependencies of some Maven package...

**Reply by laeubi** [reply]
Everything is possible but as said

1. Start small than with a whole bunch of things if you are new here
2. Sharing an example project might help to give better advice
3. In case you need dedicated support or training you can look [here](https://github.com/eclipse-tycho/tycho#getting-support)

**Reply by segevfiner** [reply]
Sadly the small thing I'm trying to do revolves around bringing in that client library and server binary to be wrapped by the plugin. As you say, I think will have to create a sample project that mocks this entire setup that I'm trying to create for anyone to help getting it working at this point. Thanks for the help up to now! BTW
